### PR TITLE
LibIMAP+Userland: Convert LibIMAP::Client to the Serenity Stream APIs

### PR DESCRIPTION
**Pull request by sin-ack**
**AK: Use Error::from_errno in adopt_nonnull_own_or_enomem**

This fails to build on Lagom otherwise.

**AK: Add ByteBuffer::{must_,}get_bytes_for_writing()**

This is useful for writing new data at the end of a ByteBuffer. For
instance, with the Stream API:

    auto pending_bytes = TRY(stream.pending_bytes());
    auto receive_buffer = TRY(buffer.get_bytes_for_writing(
        pending_bytes));
    TRY(stream.read(receive_buffer));

**LibCore: Remove the SocketError class**

SocketError is a relic from the KResult days when we couldn't have a
string in the KResult type, only an errno. Now that we can use string
literals with Error, it's no longer necessary. gai_strerror is thread
safe, so we can use it here unlike strerror.

**LibCore+Userland+Tests: Convert Stream APIs to construct on heap**

As per previous discussion, it was decided that the Stream classes
should be constructed on the heap.

While I don't personally agree with this change, it does have the
benefit of avoiding Function object reconstructions due to the lambda
passed to Notifier pointing to a stale object reference. This also has
the benefit of not having to "box" objects for virtual usage, as the
objects come pre-boxed.

However, it means that we now hit the heap everytime we construct a
TCPSocket for instance, which might not be desirable.

**LibCore: Use Error::from_errno in Stream APIs**

This makes Stream APIs work with Lagom and is overall cleaner.

**test-imap: Convert test-imap to LibMain**

**LibIMAP+Userland: Convert LibIMAP::Client to the Serenity Stream APIs**

You now cannot get an unconnected LibIMAP::Client, but you can still
close it. This makes for a nicer API where we don't have a Client object
in a limbo state between being constructed and being connected.

This code still isn't as nice as it should be, as TLS::TLSv12 is still
not a Core::Stream::Socket subclass, which would allow for consolidating
most of the TLS/non-TLS code into a single implementation.

### CODE REVIEW

**Comment by sin-ack** [comment]
Drafted while I update the Stream classes to heap-allocate during construction.

File: AK/ByteBuffer.h [hunk 1]
```diff
@@ -182,6 +182,20 @@ class ByteBuffer {
         return try_ensure_capacity_slowpath(new_capacity);
     }
 
+    /// Return a span of bytes past the end of this ByteBuffer for writing.
```
Comment (creator1creeper1) [inline AK/ByteBuffer.h:185]: `///`? Isn't that unusual?

File: Userland/Libraries/LibCore/Stream.cpp [hunk 2]
```diff
@@ -90,14 +90,14 @@ ErrorOr<off_t> SeekableStream::size()
     return seek_result.value();
 }
 
-ErrorOr<File> File::open(StringView const& filename, OpenMode mode, mode_t permissions)
+ErrorOr<NonnullOwnPtr<File>> File::open(StringView const& filename, OpenMode mode, mode_t permissions)
 {
-    File file { mode };
-    TRY(file.open_path(filename, permissions));
+    auto file = TRY(adopt_nonnull_own_or_enomem(new File(mode)));
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Stream.cpp:95]: If you want to use `adopt_nonnull_own_or_enomem`, you should use `new (nothrow)`

File: Userland/Libraries/LibCore/Stream.cpp [hunk 3]
```diff
@@ -108,8 +108,8 @@ ErrorOr<File> File::adopt_fd(int fd, OpenMode mode)
         return Error::from_errno(EINVAL);
     }
 
-    File file { mode };
-    file.m_fd = fd;
+    auto file = TRY(adopt_nonnull_own_or_enomem(new File(mode)));
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Stream.cpp:111]: See above.

File: Userland/Libraries/LibCore/Stream.cpp [hunk 4]
```diff
@@ -445,38 +445,38 @@ void PosixSocketHelper::setup_notifier()
...
 {
     auto ip_address = TRY(resolve_host(host, SocketType::Stream));
     return connect(SocketAddress { ip_address, port });
 }
 
-ErrorOr<TCPSocket> TCPSocket::connect(SocketAddress const& address)
+ErrorOr<NonnullOwnPtr<TCPSocket>> TCPSocket::connect(SocketAddress const& address)
 {
-    TCPSocket socket;
+    auto socket = TRY(adopt_nonnull_own_or_enomem(new TCPSocket()));
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Stream.cpp:456]: See above.

File: Userland/Libraries/LibCore/Stream.cpp [hunk 5]
```diff
@@ -445,38 +445,38 @@ void PosixSocketHelper::setup_notifier()
...
+ErrorOr<NonnullOwnPtr<TCPSocket>> TCPSocket::adopt_fd(int fd)
 {
     if (fd < 0) {
         return Error::from_errno(EBADF);
     }
 
-    TCPSocket socket;
-    socket.m_helper.set_fd(fd);
-    socket.setup_notifier();
+    auto socket = TRY(adopt_nonnull_own_or_enomem(new TCPSocket()));
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Stream.cpp:477]: See above.

File: Userland/Libraries/LibCore/Stream.cpp [hunk 6]
```diff
@@ -495,43 +495,43 @@ ErrorOr<size_t> PosixSocketHelper::pending_bytes() const
...
 {
     auto ip_address = TRY(resolve_host(host, SocketType::Datagram));
     return connect(SocketAddress { ip_address, port });
 }
 
-ErrorOr<UDPSocket> UDPSocket::connect(SocketAddress const& address)
+ErrorOr<NonnullOwnPtr<UDPSocket>> UDPSocket::connect(SocketAddress const& address)
 {
-    UDPSocket socket;
+    auto socket = TRY(adopt_nonnull_own_or_enomem(new UDPSocket()));
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Stream.cpp:506]: See above.

File: Userland/Libraries/LibCore/Stream.cpp [hunk 7]
```diff
@@ -495,43 +495,43 @@ ErrorOr<size_t> PosixSocketHelper::pending_bytes() const
...
-    socket.setup_notifier();
+    socket->setup_notifier();
     return socket;
 }
 
-ErrorOr<LocalSocket> LocalSocket::connect(String const& path)
+ErrorOr<NonnullOwnPtr<LocalSocket>> LocalSocket::connect(String const& path)
 {
-    LocalSocket socket;
+    auto socket = TRY(adopt_nonnull_own_or_enomem(new LocalSocket()));
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Stream.cpp:523]: See above.

File: Tests/LibCore/TestLibCoreStream.cpp [hunk 8]
```diff
@@ -300,19 +300,25 @@ TEST_CASE(local_socket_read)
     auto local_server = Core::LocalServer::construct();
     EXPECT(local_server->listen("/tmp/test-socket"));
 
+    Atomic<unsigned> should_exit_condition { 0 };
+
     local_server->on_accept = [&](NonnullRefPtr<Core::LocalSocket> server_socket) {
         EXPECT(server_socket->write(sent_data));
 
+        while (should_exit_condition.load(AK::memory_order_acquire) != 0xDEADBEEF) {
```
Comment (creator1creeper1) [inline Tests/LibCore/TestLibCoreStream.cpp:308]: `#define LOCAL_SERVER_SHOULD_EXIT 0xDEADBEEF` ?

File: Tests/LibCore/TestLibCoreStream.cpp [hunk 9]
```diff
@@ -300,19 +300,25 @@ TEST_CASE(local_socket_read)
...
+            // Let's snooze a little until the condition becomes what we expect.
+            usleep(10000);
```
Comment (creator1creeper1) [inline Tests/LibCore/TestLibCoreStream.cpp:310]: I guess this works and fixes the CI, but the whole commit feels a little hackish to me.

File: Userland/Applications/Mail/MailWidget.cpp [hunk 10]
```diff
@@ -126,12 +126,15 @@ bool MailWidget::connect_and_login()
...
-        GUI::MessageBox::show_error(window(), String::formatted("Failed to connect to '{}:{}' over {}.", server, port, tls ? "TLS" : "Plaintext"));
+    auto maybe_imap_client = tls ? IMAP::Client::connect_tls(server, port) : IMAP::Client::connect_plaintext(server, port);
+    if (maybe_imap_client.is_error()) {
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to connect to '{}:{}' over {}: {}", server, port, tls ? "TLS" : "Plaintext", maybe_imap_client.error()));
         return false;
     }
+    m_imap_client = maybe_imap_client.release_value();
+
+    auto connection_promise = m_imap_client->connection_promise();
+    VERIFY(connection_promise);
```
Comment (creator1creeper1) [inline Userland/Applications/Mail/MailWidget.cpp:137]: This isn't very expressive... You're not even `VERIFY()`ing a promise, you're verifying that a pointer to an empty promise isn't null.

File: Userland/Libraries/LibIMAP/Client.cpp [hunk 11]
```diff
@@ -460,4 +502,10 @@ void Client::close()
         m_socket->close();
     }
 }
+
+bool Client::is_open()
+{
+    return m_tls ? m_tls_socket->is_open() : m_socket->is_open();
```
Comment (creator1creeper1) [inline Userland/Libraries/LibIMAP/Client.cpp:508]: To be honest, this looks like it should be a `Variant<TLS::TLSv12, Core::Stream::Socket>`.

**Review by creator1creeper1 — CHANGES_REQUESTED** [review]

File: AK/ByteBuffer.h (hunk 1 above)
Comment (sin-ack) [reply AK/ByteBuffer.h:185]: It's kind of a preparation for API documentation in the future. The `Core::Stream` classes also have similar comments.

File: Tests/LibCore/TestLibCoreStream.cpp (hunk 8 above)
Comment (sin-ack) [reply Tests/LibCore/TestLibCoreStream.cpp:308]: Dropped this commit in favor of making the `local_socket_*` tests less thread-y in another PR.

File: Tests/LibCore/TestLibCoreStream.cpp (hunk 9 above)
Comment (sin-ack) [reply Tests/LibCore/TestLibCoreStream.cpp:310]: Dropped this commit in favor of making the `local_socket_*` tests less thread-y in another PR.

File: Userland/Applications/Mail/MailWidget.cpp (hunk 10 above)
Comment (sin-ack) [reply Userland/Applications/Mail/MailWidget.cpp:137]: Made it a little clearer.

File: Userland/Libraries/LibIMAP/Client.cpp (hunk 11 above)
Comment (sin-ack) [reply Userland/Libraries/LibIMAP/Client.cpp:508]: Well, it should really just be a `NonnullOwnPtr<Core::Stream::Socket>`, but `TLS::TLSv12` isn't a `Core::Stream::Socket` subclass yet. In the future it should all be transparent, so this is just a transitionary period until that becomes possible.

File: AK/ByteBuffer.h (hunk 1 above)
Comment (creator1creeper1) [reply AK/ByteBuffer.h:185]: 👍

**Review by creator1creeper1 — APPROVED** [review]

**Pull request merged**